File: src/commands/status.ts
import { parseArgs } from 'node:util';

import { Ticket } from '../lib/ticket.ts';
import { isValidStatus } from '../lib/types.ts';
import { exitWithError, VALID_STATUSES } from './utils.ts';

export async function cmdStatusUpdate(id: string, status: string): Promise<void> {
  if (!isValidStatus(status)) {
    exitWithError(
      `Error: invalid status '${status}'. Must be one of: ${VALID_STATUSES.join(', ')}`
    );
  }

  const ticket = await Ticket.find(id);
  await ticket.updateField('status', status);
  console.log(`Updated ${ticket.id} -> ${status}`);
}

export async function cmdStart(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 1) {
    exitWithError('Usage: tk start <id>');
  }

  return cmdStatusUpdate(positionals[0], 'new');
}

export async function cmdClose(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 1) {
    exitWithError('Usage: tk close <id>');
  }

  return cmdStatusUpdate(positionals[0], 'complete');
}

export async function cmdReopen(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 1) {
    exitWithError('Usage: tk reopen <id>');
  }

  return cmdStatusUpdate(positionals[0], 'new');
}

export async function cmdStatus(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 2) {
    exitWithError(`Usage: tk status <id> <status>\nValid statuses: ${VALID_STATUSES.join(', ')}`);
  }

  const [id, status] = positionals;
  return cmdStatusUpdate(id, status);
}

File: src/commands/status.rs
use serde_json::json;

use super::CommandOutput;
use crate::cli::OutputOptions;
use crate::error::{JanusError, Result};
use crate::ticket::Ticket;
use crate::types::TicketStatus;

/// Update a ticket's status
async fn update_status(id: &str, new_status: TicketStatus, output: OutputOptions) -> Result<()> {
    update_status_with_summary(id, new_status, None, output).await
}

/// Update a ticket's status with an optional completion summary
async fn update_status_with_summary(
    id: &str,
    new_status: TicketStatus,
    summary: Option<&str>,
    output: OutputOptions,
) -> Result<()> {
    let ticket = Ticket::find(id).await?;

    // Use the domain method that handles status updates and event logging
    ticket.update_status(new_status, summary)?;

    CommandOutput::new(json!({
        "id": ticket.id,
        "action": "status_changed",
        "new_status": new_status.to_string(),
    }))
    .with_text(format!("Updated {} -> {}", ticket.id, new_status))
    .print(output)
}

/// Set a ticket's status to "in_progress" (start working on it)
pub async fn cmd_start(id: &str, output: OutputOptions) -> Result<()> {
    update_status(id, TicketStatus::InProgress, output).await
}

/// Set a ticket's status to "complete" or "cancelled"
///
/// Requires either a summary or explicit --no-summary flag.
pub async fn cmd_close(
    id: &str,
    summary: Option<&str>,
    no_summary: bool,
    cancel: bool,
    output: OutputOptions,
) -> Result<()> {
    // Require either --summary or --no-summary
    if summary.is_none() && !no_summary {
        return Err(JanusError::SummaryRequired);
    }

    let new_status = if cancel {
        TicketStatus::Cancelled
    } else {
        TicketStatus::Complete
    };

    update_status_with_summary(id, new_status, summary, output).await
}

/// Reopen a ticket (set status back to "new")
pub async fn cmd_reopen(id: &str, output: OutputOptions) -> Result<()> {
    update_status(id, TicketStatus::New, output).await
}

/// Set a ticket's status to an arbitrary value
pub async fn cmd_status(id: &str, status: TicketStatus, output: OutputOptions) -> Result<()> {
    update_status(id, status, output).await
}
